### DISCUSSION TITLE
Support PTR records in cloaking rules?

### DISCUSSION THREAD

**Discussion by matoro** [DNSCrypt/dnscrypt-proxy#1939] [Ideas]
Right now, according to this line, the cloaking rules ignore requests for record types other than A/AAAA:  https://github.com/DNSCrypt/dnscrypt-proxy/blob/master/dnscrypt-proxy/plugin_cloak.go#L104

My network is static-IP'd, so I use cloaking rules for hardcoding names on my LAN.  However, I don't have any way to add in reverse records.

Could there be a simple syntax added to the cloaking plugin to support PTR records?  Thank you for all your hard work!

**Comment by jedisct1** [comment]
Maybe we don't even need a new syntax, and can synthesize PTR records for all entries that have an IP address?

The flip side is that is will double the amount of required memory, and some people have huge cloaking lists. 
So, maybe we also need a switch to turn this on or off.

**Reply by matoro** [reply]
That could be tricky with multiple names that point to the same address (pick the first?)

I do also use the wildcard cloaking feature (skip those?)

**Comment by ianbashford** [comment]
@matoro 
I've started to build up a PR for this feature.  https://github.com/DNSCrypt/dnscrypt-proxy/pull/1958/files  
If you have the ability to build this then you can test the 'work in progress'  version and see if it works for you.

It should function correctly, but I do still need to build in the switch to make this optional for those with large cloaking files.

To build, simply checking out that branch, and building with something like `go build -ldflags="-s -w" -v`  will suffice to get you a working copy to test with.

**Reply by matoro** [reply]
Wow, that's great!  Let me try it out right now!

**Reply by matoro** [reply]
@ianbashford hmm, seemed to crash while parsing my cloaking rules:

```
[2021-12-11 15:35:37] [NOTICE] Loading the set of cloaking rules from [/etc/dnscrypt-proxy/cloaking-rules.txt]
[2021-12-11 15:35:37] [FATAL] Syntax error in block rules at pattern 22
```

**Reply by ianbashford** [reply]
Could you give me a clue as to what the rule looks like there?

**Reply by matoro** [reply]
Does the 22 correspond to the line number?  The rules look something like this:

```
=hostone.mydomain.com    fd02::1
=hostone.mydomain.com    192.168.2.1

=hosttwo.mydomain.com    fd02::2
=hosttwo.mydomain.com    192.168.2.2

...

*.mydomain.com    fd02::1
*.mydomain.com    192.168.2.1

=anotherhost.mydomain.com    fd03::2
=anotherhost.mydomain.com    192.168.2.1
=cloud.mydomain.com    external.cname.alias.com
=ext.mydomain.com    1.2.3.4
```

Then a bunch of custom tracking domain blocks in the form:
```
*.addomain.com    0.0.0.0
````

Line 22 corresponds to this line:
```
=anotherhost.mydomain.com    fd03::2
```

**Reply by ianbashford** [reply]
OK I think I've fixed it -- it works now with the entries you have given me above.
I've just pushed it and it's going through CI now... 
Note that I've added the config line as well, so you'll need to add `cloak_ptr = true`  into the main config file as well

**Reply by matoro** [reply]
It works!  Thank you so much, you rock man!

**Reply by matoro** [reply]
Oh, I have got one more question - is there a way to add PTR records without a corresponding A/AAAA one?  For example, as you can see I give all my devices ULA IPv6 addresses, but they also have globally addressable ones.  I'd like the global addresses to reverse-resolve to the correct names, but I don't want the AAAA records for the hosts to resolve to the global addresses, as all internal communications should use the ULA ones.  Is that possible?
